### PR TITLE
(WIP) Sparse models: S-LASSO and TV-l1

### PR DESCRIPTION
**Pull request by dohmatob**
- Supports TV-l1 and S-LASSO priors
- Supports logistic and squared losses
- Has cross validation
- Can automatically select alpha by CV (+ automatic computation of useful alpha ranges for the CV)
- Warning: User must supply l1_ratio

### CODE REVIEW

**Comment by dohmatob** [comment]
Hi,

Discusion:
Is there any plan to merge this PR anytime soon ? Concerning the doc (sphinx), there're probably one or two things to add, but otherwise the code is mergeable and usable rightaway.

**Comment by GaelVaroquaux** [comment]
> Is there any plan to merge this PR anytime soon ? Concerning the doc
> (sphinx), there're probably one or two things to add, but otherwise the
> code is mergeable and usable rightaway.

Can you paste the figure output by the examples so that we have a look?

Also, indeed there is some work to be done on the docs.

**Comment by dohmatob** [comment]
oasis_tvl1_pred
oasis_tvl1_map
oasis_sl_pred
oasis_sl_map
haxb_ytvl1
haxby_sl
poldrack_tvl1
poldrack_sl

**Comment by GaelVaroquaux** [comment]
Hi @dohmatob : 

The images look pretty good in general. Comments:

* On the Oasis data, could you put in the title of the plot of true and predicted age a measure of prediction error (eg MSE)

* On the Poldrack data, why is there no anatomy in the background?

**Comment by dohmatob** [comment]
On Wed, Feb 4, 2015 at 2:55 PM, Gael Varoquaux <notifications@github.com>
wrote:

> Hi @dohmatob <https://github.com/dohmatob> :
>
> The images look pretty good in general. Comments:
>
>    -
>
>    On the Oasis data, could you put in the title of the plot of true and
>    predicted age a measure of prediction error (eg MSE)
>
> Indeed

>
>    -
>
>    On the Poldrack data, why is there no anatomy in the background?
>
> OK, poldrack has no anatomic data. We'll use a template (since the data is
supposed to be already normalized). Is there any packaged / fetchable
template with /  nilearn ?

>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/219#issuecomment-72857596>.
>



-- 
DED

**Comment by mrahim** [comment]
> Is there any packaged / fetchable template with / nilearn ?

You can remove `background_img` from `plot_stat_map`

figure_1

**Comment by dohmatob** [comment]
Oops! The bg is a zombie code snipette from my OHBM submission deadline
rumble (in the past, the thumbs were without bg). Thanks for pointed out!
Any news about availability templates in nilearn anybody ?
On Wed, Feb 4, 2015 at 4:28 PM, mrahim <notifications@github.com> wrote:

> Is there any packaged / fetchable template with / nilearn ?
>
> You can remove background_img from plot_stat_map
>
> [image: figure_1]
> <https://cloud.githubusercontent.com/assets/8612476/6043179/cc13308c-ac8a-11e4-94e2-03471d5753a8.png>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/219#issuecomment-72874708>.
>



-- 
DED

**Comment by dohmatob** [comment]
Oops! @mrahim ignore my last msg; off-topic. Thanks for the trick.

On Wed, Feb 4, 2015 at 5:34 PM, Elvis Dohmatob <gmdopp@gmail.com> wrote:

> Oops! The bg is a zombie code snipette from my OHBM submission deadline
> rumble (in the past, the thumbs were without bg). Thanks for pointed out!
> Any news about availability templates in nilearn anybody ?
>
> On Wed, Feb 4, 2015 at 4:28 PM, mrahim <notifications@github.com> wrote:
>
>> Is there any packaged / fetchable template with / nilearn ?
>>
>> You can remove background_img from plot_stat_map
>>
>> [image: figure_1]
>> <https://cloud.githubusercontent.com/assets/8612476/6043179/cc13308c-ac8a-11e4-94e2-03471d5753a8.png>
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/nilearn/nilearn/pull/219#issuecomment-72874708>.
>>
>
>
>
> --
> DED
>



-- 
DED

**Comment by dohmatob** [comment]
poldrack_tvl1
poldrack_sl

**Comment by dohmatob** [comment]
oasis_tvl1_pred
oasis_sl_curve

**Comment by GaelVaroquaux** [comment]
OK. The figures look good. Good job.

We need narrative documentation. We'll talk about where to integrate it tomorrow.

**Comment by lesteve** [comment]
Before I forget, the examples will need to be moved to their relevant subfolders, like examples/decoding for example.

Also for plot_haxby_space_net.py and plot_oasis_vbm_space_net.py, the examples they were based on have changed cosmetically but significantly compared to when this work started. For obvious uniformity reasons, it would be very nice if your examples could be updated accordingly. 

Stuff I can think of off the top my head, using index_img, some uniformity with dataset naming and using variables with filename in them to make it more obvious what dataset_files.cmaps[0] is for example. It is quite easy to figure out by looking at an up-to-date version of plot_haxby_simple.py and plot_oasis_vbm.py.

**Comment by dohmatob** [comment]
@lesteve: Yes definitely. Thanks for noting this. I also had this in mind.

File: nilearn/decoding/space_net.py [hunk 1]
```diff
@@ -0,0 +1,1242 @@
...
+    default 20)
+        Percentile value for ANOVA univariate feature selection. A value of
+        100 means 'keep all features'. This percentile is is expressed
+        w.r.t the volume of a standard (MNI152) brain, and so is corrected
+        at runtime to correspond to the volume of the user-supplied mask
+        (which is typically smaller).
+
+    standardize : bool, optional (default True):
+        If set, then the data (X, y) are centered to have mean zero along
+        axis 0. This is here because nearly all linear models will want
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: There is a problem here: `standardize` is given to a NiftiMasker which centers AND norms the data. Is the `normalize` parameter supposed to do something else?

File: nilearn/decoding/space_net.py [hunk 2]
```diff
@@ -0,0 +1,1242 @@
...
+    i_max = idx[0].max()
+    j_min = max(idx[1].min() - 1, 0)
+    j_max = idx[1].max()
+    k_min = max(idx[2].min() - 1, 0)
+    k_max = idx[2].max()
+    return mask[i_min:i_max + 1, j_min:j_max + 1, k_min:k_max + 1]
+
+
+def _univariate_feature_screening(
+        X, y, mask, is_classif, screening_percentile, smooth=2.):
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: Really useful function. A few suggestions:
- we should maybe put it into _utils, I would use it for the `Decoder` class
- should it be called check_univariate_screening instead?
- should we create a _utils/checkers.py files for this?
- could the `screening_percentile` argument just be called `screening`? In the classes constructors as well?
- this would integrate nicely with a `compressing` argument for clustering methods in future averaging models for example

File: nilearn/decoding/space_net.py [hunk 3]
```diff
@@ -0,0 +1,1242 @@
...
+
+    is_classif: bool
+        Flag telling whether the learning task is classification or regression.
+
+    screening_percentile : float in the closed interval [0., 100.]
+        Only the `screening_percentile * 100" percent most import voxels will
+        be retained.
+
+    smooth : float, optional (default 2.)
+        FWHM for isotropically smoothing the data X before F-testing. A value
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: `smooth` is called `smoothing_fwhm` in the NiftiMaskers

File: nilearn/decoding/space_net.py (hunk 1 above)
Comment (dohmatob) [inline nilearn/decoding/space_net.py]: The doc for `normalize` is zombie code. This parameter is not used anymore. Removed.

File: nilearn/decoding/space_net.py [hunk 4]
```diff
@@ -0,0 +1,1242 @@
...
+MNI152_BRAIN_VOLUME = 1827243.
+
+
+def _get_mask_volume(mask):
+    """Computes the volume of a brain mask in mm^3
+
+    Parameters
+    ----------
+    mask : nibabel image object
+        Input image whose voxel dimensions are to be computed.
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: `mask` should be `mask_img`

File: nilearn/decoding/space_net.py [hunk 5]
```diff
@@ -0,0 +1,1242 @@
...
+    eps : float, optional (default 1e-3)
+        Length of the path. For example, ``eps=1e-3`` means that
+        ``alpha_min / alpha_max = 1e-3``
+
+    alpha_min : float, optional (default None)
+        Minimum value of alpha to consider. This is mutually exclusive with the
+        `eps` parameter.
+
+    mask : filename, niimg, NiftiMasker instance, optional default None)
+        Mask to be used on data. If an instance of masker is passed,
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: `mask` should be `mask_img`

File: nilearn/decoding/space_net.py [hunk 6]
```diff
@@ -0,0 +1,1242 @@
...
+        if is_classif:
+            w_ = np.append(w_, best_w[-1])
+            w_[:-1][support] = best_w[:-1]
+        else:
+            w_[support] = best_w
+        best_w = w_
+
+    if len(best_w) == n_features:
+        if Xmean is None:
+            Xmean = np.zeros(n_features)
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: pep8 nitpicking: Xmean to X_mean. This exception to pep8 is Xmas.

File: nilearn/decoding/space_net.py [hunk 7]
```diff
@@ -0,0 +1,1242 @@
...
+            if isinstance(l1_ratios, numbers.Number):
+                l1_ratios = [l1_ratios]
+            for l1_ratio in l1_ratios:
+                if not 0 <= l1_ratio <= 1.:
+                    raise ValueError(
+                        "l1_ratio must be in the interval [0, 1]; got %g" % (
+                            l1_ratio))
+                elif l1_ratio == 0. or l1_ratio == 1.:
+                    warnings.warn(
+                        ("Specified l1_ratio = %g. It's adived to only "
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: It's advised to only

File: nilearn/decoding/space_net.py [hunk 8]
```diff
@@ -0,0 +1,1242 @@
...
+                        self.SUPPORTED_LOSSES) > 2 else "",
+                    self.SUPPORTED_LOSSES[-1], self.loss))
+        if not self.loss is None and not self.is_classif and (
+            self.loss == "logistic"):
+            raise ValueError(
+                ("'logistic' loss is only available for classification "
+                 "problems."))
+
+    def _set_coef_and_intercept(self, w):
+        """Sets the loadings vector (coef) and the intercept of the fitted
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: Set the loadings vector

File: nilearn/decoding/space_net.py [hunk 9]
```diff
@@ -0,0 +1,1242 @@
...
+            self.intercept_ = self.w_[:, -1]
+        else:
+            self._set_intercept(self.Xmean_, self.ymean_, self.Xstd_)
+
+    def fit(self, X, y):
+        """Fit the learner
+
+        Parameters
+        ----------
+        X : list of filenames or NiImages of length n_samples, or 2D array of
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: Does it really work with np arrays? Doesn't TV require an affine (and therefore) a Niimage? Doesn't the
 _univariate_feature_screening also require an affine to estimate the brain volume? Anyway the fit method appears to require a NiftiMasker that fit_transforms the X argument. Which would fail if X is an array.

File: nilearn/decoding/space_net.py [hunk 10]
```diff
@@ -0,0 +1,1242 @@
...
+                                       target_affine=self.target_affine,
+                                       target_shape=self.target_shape,
+                                       standardize=self.standardize,
+                                       low_pass=self.low_pass,
+                                       high_pass=self.high_pass,
+                                       mask_strategy='epi', t_r=self.t_r,
+                                       memory=self.memory_)
+        X = self.masker_.fit_transform(X)
+        self.Xmean_ = X.mean(axis=0)
+        self.Xstd_ = X.std(axis=0)
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: pep8

File: nilearn/decoding/space_net.py [hunk 11]
```diff
@@ -0,0 +1,1242 @@
...
+             y_train_mean, (cls, fold)) in Parallel(
+            n_jobs=self.n_jobs, verbose=2 * self.verbose)(
+            delayed(self.memory_.cache(path_scores))(
+                solver, X, y[:, cls] if n_problems > 1 else y, self.mask_,
+                alphas, l1_ratios, self.cv_[fold][0], self.cv_[fold][1],
+                solver_params, n_alphas=self.n_alphas, eps=self.eps,
+                is_classif=self.loss == "logistic", key=(cls, fold),
+                debias=self.debias, verbose=self.verbose,
+                screening_percentile=self.screening_percentile_
+                ) for cls in xrange(n_problems) for fold in xrange(n_folds)):
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: I think it is better to use range rather than xrange

File: nilearn/decoding/space_net.py [hunk 12]
```diff
@@ -0,0 +1,1242 @@
...
+
+    loss: string, optional (default "logistic"):
+        Loss to use in the classification problems. Must be one of "mse" and
+        "logistic".
+
+    eps : float, optional (default 1e-3)
+        Length of the path. For example, ``eps=1e-3`` means that
+        ``alpha_min / alpha_max = 1e-3``
+
+    mask : filename, niimg, NiftiMasker instance, optional default None)
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: mask_img

File: nilearn/decoding/space_net.py [hunk 13]
```diff
@@ -0,0 +1,1242 @@
...
+
+    n_alphas : int, optional (default 10).
+        Generate this number of alphas per regularization path.
+        This parameter is mutually exclusive with the `alphas` parameter.
+
+    eps : float, optional (default 1e-3)
+        Length of the path. For example, ``eps=1e-3`` means that
+        ``alpha_min / alpha_max = 1e-3``
+
+    mask : filename, niimg, NiftiMasker instance, optional default None)
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: mask_img

File: nilearn/decoding/space_net.py (hunk 11 above)
Comment (dohmatob) [inline nilearn/decoding/space_net.py]: Yes indeed, xrange will become range in py3, blablabla

File: nilearn/decoding/space_net.py [hunk 14]
```diff
@@ -0,0 +1,1230 @@
...
+    i_max = idx[0].max()
+    j_min = max(idx[1].min() - 1, 0)
+    j_max = idx[1].max()
+    k_min = max(idx[2].min() - 1, 0)
+    k_max = idx[2].max()
+    return mask[i_min:i_max + 1, j_min:j_max + 1, k_min:k_max + 1]
+
+
+def _univariate_feature_screening(
+        X, y, mask, is_classif, screening_percentile, smoothing_fwhm=2.):
```
Comment (schwarty) [inline nilearn/decoding/space_net.py]: Really useful function. A few suggestions:
- we should maybe put it into _utils, I would use it for the `Decoder` class
- should it be called check_univariate_screening instead?
- should we create a _utils/checkers.py files for this?
- could the `screening_percentile` argument just be called `screening`? In the classes constructors as well?
- this would integrate nicely with a `compressing` argument for clustering methods in future averaging models for example

File: nilearn/decoding/space_net.py (hunk 9 above)
Comment (dohmatob) [inline nilearn/decoding/space_net.py]: Deceptive docstring, fixing...

File: nilearn/decoding/space_net.py (hunk 14 above)
Comment (dohmatob) [inline nilearn/decoding/space_net.py]: The function is not meant to become a tool. It's just a piece of inline code which got factored out for software-engineering reasons.

**Comment by AlexandreAbraham** [comment]
@dohmatob @schwarty @GaelVaroquaux What is the status on this PR?

**Comment by dohmatob** [comment]
The PR is pretty much complete. The only thing left is to make the examples
integrate into the docs (this has been started). I'll allocate some time to
address it next week..

On Wed, Jun 3, 2015 at 1:27 PM, Alexandre Abraham <notifications@github.com>
wrote:

> @dohmatob <https://github.com/dohmatob> @schwarty
> <https://github.com/schwarty> @GaelVaroquaux
> <https://github.com/GaelVaroquaux> What is the status on this PR?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/219#issuecomment-108314717>.
>



-- 
DED

File: doc/sphinxext/gen_rst.py [hunk 15]
```diff
@@ -671,7 +675,7 @@ def generate_file_rst(fname, target_dir, src_dir, root_dir, plot_gallery):
     time_elapsed = 0
     time_m = 0
     time_s = 0
-    if plot_gallery and fname.startswith('plot'):
+    if plot_gallery and demo_regexp.search(fname):
```
Comment (GaelVaroquaux) [inline doc/sphinxext/gen_rst.py]: You cannot have changes to gen_rst in this PR: they need to go to sphinx-gallery.

**Comment by GaelVaroquaux** [comment]
> Closed #219.

219 closed. Sniff! That was an old friend.

Looking forward to see its daugther.

**Comment by dohmatob** [comment]
On Tue, Jul 14, 2015 at 9:34 AM, Gael Varoquaux <notifications@github.com>
wrote:

> > Closed #219.
>
> 219 closed. Sniff! That was an old friend.
>
[Chuckles] true true

>
> Looking forward to see its daugther.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/219#issuecomment-121153206>.
>



-- 
DED
